Support for any_discontinuous_space (for #133)

TeranIvy opened the pull request:
#### PR to keep track of changes for `any_discontinuous_space` support (issue #133)

- Extend optimisation rules for discontinuous spaces to  `any_discontinuous_space`  (e.g. no colouring, loops only over owned cells).
- Updated `DynamoLoopFuseTrans` logic and separated checks into `_validate` method. Also updated documentation.
- Updated stub generation and created related issue #461.
- Modified LMA and CMA operator tests. Created issue #465 to remove duplicate LMA tests.
- Modified and extended multigrid tests and code examples.
- Added and modified stencil examples for `any_discontinuous_space`.
- Modified LFRic examples  (`dg_matrix_vector_kernel_mod.F90` and multigrid kernels) to resemble the usage for `any_discontinuous_space` in LFRic trunk. 
 Tom Melvin **tested the relevant LFRic branch** (see [LFRic ticket 1903](https://code.metoffice.gov.uk/trac/lfric/ticket/1903)) and confirmed correctness of the `gungho` executable.
- Modified and extend documentation.
- General updates:
  - Added compilation testing for modifed kernels and test files;
  - Updated Copyrights and tidied up the code;
  - Replaced several occurences of `GH_WRITE` access for continuous fields with `GH_INC` (see #138) which led to some test modifications.

The full list of changes is here: [133_list_of_changes.txt](https://github.com/stfc/PSyclone/files/3513469/133_list_of_changes.txt)

_Note:_ LFRic trunk was updated to support `any_discontinuous_space` metadata in [LFRic ticket 1906](https://code.metoffice.gov.uk/trac/lfric/ticket/1906).

TeranIvy commented:
Coverage update does not seem to work so I will temporarily close the PR. Hopefully reopening will trigger the coverage.

TeranIvy commented:
Before proceeding with changes to tests, I will need to change metadata from `ANY_D_SPACE_*` to `ANY_DISCONTINUOUS_SPACE_*` as agreed in today's design discussion (see [comment](https://github.com/stfc/PSyclone/issues/133#issuecomment-520891616)).

TeranIvy commented:
All tests pass, including compilation with Intel 17.0.1 and Gnu 6.1.0. Ready for the first review from @rupertford.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -59,24 +59,45 @@
+     "any_discontinuous_space_9", "any_discontinuous_space_10"]
+
+# Valid discontinuous FS names (for optimisation purposes)
+VALID_DISCONTINUOUS_FUNCTION_SPACE_NAMES = DISCONTINUOUS_FUNCTION_SPACES + \
```
I know you are inheriting this problem, but I don't like the mixing of *FUNCTION_SPACE_NAMES and *FUNCTION_SPACES. IMHO they are all names so should have _NAMES at the end and are all VALID so don't need VALID at the start. Fixing this is outside of the remit of this PR but to make things a little better perhaps VALID_DISCONTINUOUS_FUNCTION_SPACE_NAMES could be called ALL_DISCONTINUOUS_FUNCTION_SPACE_NAMES and VALID_ANY_DISCONTINUOUS_FUNCTION_SPACE_NAMES could be called ANY_DISCONTINUOUS_FUNCTION_SPACE_NAMES?

TeranIvy replied:
What is the specific remark about this and is there a specific issue to address this? I couldn't find any so I am happy to open a new issue so we don't forget about this.
I'd rather come to an agreement in this issue to not have to modify optimisation scripts again but I am also happy to leave it for another issue as long as there is a comment in the code as a reminder :-)

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -174,21 +171,10 @@ def check_intergrid(node):
-        >>> new_schedule.view()
+    ''' Provides a generic loop-fuse transformation to two Nodes in the
+    PSyIR of a Schedule after performing validity checks for the supplied
+    Nodes. Examples are given in descriptions of children classes
```
Change to "in the descriptions of any children classes." and remove the reforence to  the children classes as parent classes should not know what they children classes are.

TeranIvy replied:
Done.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -237,52 +222,75 @@ def _validate(self, node1, node2):
                                       "next to each other")
-
-        # Check iteration space is the same
+        # Check that iteration space is the same
```
... that the interation ...

TeranIvy replied:
Done.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -237,52 +222,75 @@ def _validate(self, node1, node2):
+        ''' Fuses two Loops represented by `psyclone.psyGen.Node` objects
+        after performing validity checks.
+
+        :param node1: the first Node we are checking
```
Please write in the 3rd person e.g. "the first Node that is being checked."

TeranIvy replied:
Done.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -237,52 +222,75 @@ def _validate(self, node1, node2):
+
+        :param node1: the first Node we are checking
+        :type node1: :py:class:`psyclone.psyGen.Node`
+        :param node2: the second Node we are checking
```
3rd person here too please.

TeranIvy replied:
Done.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
+                        "optional argument to 'True'.".format(self.name))
+                # 2.3.2) Check whether specific function spaces are the
+                # same. If they are not, the loop fusion is still possible
+                # but only when both function spaces are discontinuous
```
I don't believe that discontinuous means we can fuse (or does it?)

TeranIvy replied:
See the above comment about user-supplied kernels looping over cells (safe to fuse) and not dofs (not safe to fuse).

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -237,52 +222,75 @@ def _validate(self, node1, node2):
-        type) before calling the :py:meth:`LoopFuseTrans.apply` method of the
-        :py:class:`base class <LoopFuseTrans>` in order to fuse two
-        GOcean loops. '''
+    ''' GOcean API application of the :py:class:`base class <LoopFuseTrans>`
```
Done

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -294,20 +302,29 @@ def name(self):
     def apply(self, node1, node2):
-        '''Fuse the two GOcean loops represented by :py:obj:`node1` and
-        :py:obj:`node2`.
+        ''' Fuses two `psyclone.gocean1p0.GOLoop` Loops after performing
```
Done

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
+        same_space = self._same_space_flag
+
+        # Call the parent class validation first
+        super(DynamoLoopFuseTrans, self)._validate(node1, node2)
```
It is! I missed the duplication. Removed the one in line 418.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -294,20 +302,29 @@ def name(self):
+        :rtype: (:py:class:`psyclone.psyGen.Schedule`, \
+                 :py:class:`psyclone.undoredo.Memento`)
+
+        :raises TransformationError: if the supplied Loops are over \
```
Done.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -324,13 +341,35 @@ def apply(self, node1, node2):
-        to fuse two Dynamo loops.
+    ''' Dynamo0.3 API application of the
+    :py:class:`base class <LoopFuseTrans>` in order to fuse two Dynamo
+    Loops after performing validity checks. For example:
```
Done everywhere in modified/added comments.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
                                      an inter-grid kernel.
+        :raises TransformationError: if one or both function spaces have \
+                                     invalid names.
+        :raises TransformationError: if the `same_space`flag was set, but \
```
Done

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +666,25 @@ def __init__(self, arg_type):
+               self._access_type == AccessType.INC:
+                raise ParseError(
+                    "In the Dynamo0.3 API a field on any_discontinuous_space "
+                    "cannot have 'gh_inc' access because it is treated "
```
Yes. Changed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3342,6 +3379,11 @@ def basis_first_dim_value(function_space):
             first_dim = "3"
         else:
+            # It is not possible to determine explicitly the first basis
+            # function array dimension from the metadata for any_space and
```
Changed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3342,6 +3379,11 @@ def basis_first_dim_value(function_space):
+            # It is not possible to determine explicitly the first basis
+            # function array dimension from the metadata for any_space and
+            # any_discontinuous_space. This information needs to be passed
+            # from the PSy layer to the kernels, which will be enabled
```
Changed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3384,6 +3430,11 @@ def diff_basis_first_dim_value(function_space):
+            # It is not possible to determine explicitly the first
+            # differential basis function array dimension from the metadata
+            # for any_space and any_discontinuous_space. This information
+            # needs to be passed from the PSy layer to the kernels, which
```
Changed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -59,24 +59,45 @@
 
-VALID_FUNCTION_SPACE_NAMES = VALID_FUNCTION_SPACES + VALID_ANY_SPACE_NAMES
+# Valid any_discontinuous_space metadata (general FS known to be discontinuous)
+VALID_ANY_DISCONTINUOUS_SPACE_NAMES = \
```
Happy with as list comprehension, thanks for the sugestion. I will leave comprehension for `any_space` for #455.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -1397,7 +1397,6 @@ def walk(self, my_type):
         an instance of 'my_type', which is either a single class or a tuple
         of classes. In the latter case all nodes are returned that are
         instances of any classes in the tuple.
```
I don't remember why I removed that. I restored it and also added a space for the return value to be consisent with the guidelines (and added spaves accordingly in other modified docstrings).

TeranIvy commented on doc/thebigpicture.mm:
```diff
@@ -321,7 +321,7 @@
 <icon BUILTIN="button_cancel"/>
 </node>
-<node COLOR="#000000" CREATED="0" ID="ID_1910239446" MODIFIED="0" TEXT="any_d_space">
+<node COLOR="#000000" CREATED="0" ID="ID_1910239446" MODIFIED="0" TEXT="any_discontinuous_space">
```
Missed that! Changed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -59,24 +59,45 @@
-# First section : Parser specialisations and classes
-
-# Function spaces (FS)
+# --------------------------------------------------------------------------- #
```
My intention here was to make "headers" for different functionality sections and subsections as the file is very long and it can be difficult to find a specific thing. Ideally, all these headers would not be required as we would have different pieces of LFRic API functionality in more than one file (see Joerg's restructuring quesion #474).  I googled for a standard way of commenting in Python and could not find anything special so I guess this is as good way as any - I am open to suggestions how to organise and mark up comments in code!

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +666,25 @@ def __init__(self, arg_type):
+                    "In the Dynamo0.3 API a field on any_discontinuous_space "
+                    "cannot have 'gh_inc' access because it is treated "
+                    "as discontinuous")
+            # TODO (issue #138): extend for "gh_write"
```
Since we now have `ANY_DISCONTINUOUS_SPACE` I don't see the point of disallowing `GH_WRITE` for continuous spaces and not include `ANY_SPACE` in this restriction for the user-supplied kernels that iterate over cells.  `ANY_SPACE` is essentially treated as continuous and hence requires colouring when iterating over cells.
Of course, there is a question of dealing with built-ins as they loop over DoFs and not over cells. In their case using `GH_WRITE` (and `GH_READWRITE` instead of `GH_INC`) with `ANY_SPACE` makes sense. I opened the issue #471 for that.

I updated descriptions of #138 and #471 to distinguish between looping over cells and DoFs and modified TODO for `ANY_SPACE` to point to issue #471 instead.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -237,52 +222,75 @@ def _validate(self, node1, node2):
+    >>>
+    >>> from psyclone.transformations import GOceanLoopFuseTrans
+    >>> ftrans = GOceanLoopFuseTrans()
+    >>> new_schedule, memento = ftrans.apply(schedule.children[0],
```
I didn't see this level of `schedule.children` changed neither in Joerg's #454 (simplify walk) nor in Sergi's #400 for `DO` loops and I updated the branch with the master after those two PRs were merged.

TeranIvy replied:
I presume `__getitem__` in `Schedule` instead of `get_item`? I modified this but I did not modify the related transformation scripts in `examples/gocean` to this way of applying the transformation because there are quite a lot of changes already. Plus, there are a couple more scripts that apply `LoopFuseTrans` in `dynamo/eg2` which should then also be modified this way. I am happy to update the scripts in this issue or leave it for another one.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -324,13 +341,35 @@ def apply(self, node1, node2):
+    >>> from psyclone.transformations import DynamoLoopFuseTrans
+    >>> ftrans =  DynamoExtractRegionTrans()
+    >>>
+    >>> new_schedule, memento = ftrans.apply(schedule.children[0],
```
I think this is still the same as it wasn't changes in #454 or #400 (see the comment above).

TeranIvy replied:
Modified as requested and updated transformation scripts in `examples/dynamo/eg6` to reflect this way of applying the transformation. Also corrected
```
ftrans =  DynamoExtractRegionTrans()
```
to
```
ftrans =  DynamoLoopFuseTrans()
```

rupertford replied:
Thanks for fixing. You made a small mistake (```children[1]``` instead of ```schedule[1]```) but I've just updated that myself rather than request another review.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -412,27 +508,56 @@ def apply(self, node1, node2, same_space=False):
+                                "loops as the first loop has a reduction "
+                                "and the second loop reads the result of "
+                                "the reduction.".format(self.name))
+        # 6) Capture Transformation class error
```
I hope not as there wasn't a test for it anyway. Removed.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
-            # Check that we don't have an inter-grid kernel
+        from psyclone.dynamo0p3 import VALID_FUNCTION_SPACE_NAMES, \
+            VALID_DISCONTINUOUS_FUNCTION_SPACE_NAMES
+        try:
```
For the unexpected exception handler perhaps?

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
-        :py:obj:`node2`. The optional same_space flag asserts that an
-        unknown iteration space (i.e. any_space) matches the other
-        iteration space. This is set at the users own risk.
+    def _validate(self, node1, node2):
```
It is different to `apply()` having different number of arguments because `apply()` is not declared as a private method (with an underscore `_`) as is `_validate`. Apparently, having different number of arguments  in children classes' methods works for public methods but not for private. My original intention was to pass `same_space` as an argument to the `DynamoLoopFuseTrans._validate`  method because I assumed it would be possible to "overload" the functionality (different number of arguments) of the parent class. However, this failed miserably because the parent `LoopFuseTrans.apply` method complained about the positional arguments when it called `self._validate(node1, node2)`. Making `validate` a public method removed this issue but it is in disagreement with how `_validate` is applied in the rest of transformations.

TeranIvy replied:
Used setter as suggested in the previous comment and added TODO with the reference to issue #478, as having this functionality will make the setter redundant.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -412,27 +508,56 @@ def apply(self, node1, node2, same_space=False):
         except TransformationError as err:
             raise err
+
+        # 7) Capture unexpected error
```
Removed the exception and the related test.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
+                    "spaces '{1}' and '{2}' have invalid names.".
+                    format(self.name, node1_fs_name, node2_fs_name))
+            # Check whether any of the spaces is ANY_SPACE. Loop fusion over
+            # ANY_SPACE is allowed only when the 'same_space' flag is set
```
It may be simpler but I think it would be inaccurate and ambiguous when it comes to the `same_space` use. If I understood the original logic correctly, `same_space` was intented to be used only when at least one of the loops is on `ANY_SPACE` but the user knows they are same. Otherwise it does not make much sense.
Unconditional `same_space` flag may be dangerous if there is no additional check that the specific spaces are indeed the same (hence duplicating specific check space from the propsed "elif" condition). Also, it may be applied to two same specific spaces where it does not need to be used (as in the corrected test in `psyGen_test` where both spaces are `w1` and `same_space` is used on master.

TeranIvy replied:
After a careful consideration I decided to allow the use of `same_space` flag when the function spaces are not `ANY_SPACE` but have the same name. This can be useful when loop fusion is applied generically in LFRic.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -324,13 +341,35 @@ def apply(self, node1, node2):
+    def __init__(self):
+        # Define private variable for storing the value of the optional
+        # 'same_space' flag passed to the 'apply' method. Otherwise we
+        # would have to name mangle the '_validate' methods of parent
```
Used setter as suggested in the previous comment and added TODO with the reference to issue #478, as having this functionality will make the setter redundant.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -340,67 +379,124 @@ def name(self):
+                                     is unknown (`ANY_SPACE`) and the \
+                                     `same_space` optional argument is not \
+                                     set to `True`.
+        :raises TransformationError: if the Loops are over different spaces \
```
As discussed today at the Telco, this is fine as we are looping over cells in the user-supplied kernels. If we were looping over dofs than two different discontinuous spaces would have different number of dofs so it would not be safe to fuse the loops. However, looping over dofs is currenlty enabled only in built-ins which are on `ANY_SPACE` anyway.

TeranIvy commented:
I think I addressed all review comments. The branch is updated with the latest master. Implementing setters for the `same_space` flag led to the need to update some optimisation scripts in the `dynamo/examples` directory. I also corrected `README.md` and `eg12/find_kernel.py` in examples which were incorrect after renaming `matrix_vector_kernel_code` in PR #442.

Looking at the `loop_fuse.py` script in examples I realised we need to allow the `same_space` flag to be set to `True` when both spaces have the same name and are not on `ANY_SPACE`. This will guarantee the correct behaviour when the loop fusion is applied generically over all loops in an algorithm and there is no harm done if the transformation is only applied to two loops that are over the same space. Hence, I added this in the `DynamoLoopFuseTrans` and reverted the `test_node_args` (`psyGen_test.py`) to have `same_space` set to `True` in order to check whether this logic indeed works for two spaces that have the same name.

Two more things to mention:
- Loop fuse logic over discontinuous spaces is left as it is, as discussed at the today's Telco (allowed for any 2 discontinuous spaces as they loop over `ncells`, unless the further check detects that the upper loop bounds are not the same due to some other transformation),
- Underscores in tests that return `TransformationError` are not removed due to `pylint` warnings.

Ready for another review from @rupertford.

rupertford commented on src/psyclone/transformations.py:
```diff
@@ -412,99 +463,203 @@ def name(self):
+        :raises TransformationError: if one or more of the iteration spaces
+                                     is unknown (`ANY_SPACE`) and the \
+                                     `same_space` flag is not set to `True`.
+        :raises TransformationError: if the loops are over different spaces \
```
You explained that this test is only when iterating over cells. Please explain this in the description so people know the context of the test e.g. add "and the loops both iterate over cells." at the end.

TeranIvy replied:
Thanks for the suggestion. Added.

rupertford replied:
Great. There was a missing ```\``` which I've just added as it is just a very minor change.

rupertford commented:
Back to @TeranIvy. Please note that a couple of earlier conversations are still open (they appear further up the page).

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +662,26 @@ def __init__(self, arg_type):
                     "space ({0}) to have a 'gh_inc' access".
                     format(self._function_space1.lower()))
-            # TODO: extend for "gh_write"
+            if self._function_space1.lower() in \
```
Good idea, thanks. Modified the existing test as suggested, updated `test_fs_discontinuous_space_and_inc_error` in `dynamo0p3_test.py` to reflect this and removed `test_fs_any_discontinuous_space_and_inc_error` as duplicate.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +662,26 @@ def __init__(self, arg_type):
+                raise ParseError(
+                    "In the Dynamo0.3 API a field on any_discontinuous_space "
+                    "cannot have 'gh_inc' access because it is discontinuous")
+            # TODO (issue #138): extend for "gh_write"
```
Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +662,26 @@ def __init__(self, arg_type):
                     "space ({0}) to have a 'gh_readwrite' access".
                     format(self._function_space1.lower()))
-            # TODO: extend for "gh_write"
+            # TODO: extend restriction to "gh_write" for kernels that loop
```
I agree, this would be the simplest logic. Modified the `#TODO` comment to reflect this and quoted the logic in issue #471.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -639,18 +662,26 @@ def __init__(self, arg_type):
                     format(self._function_space1.lower()))
-            # TODO: extend for "gh_write"
+            # TODO: extend restriction to "gh_write" for kernels that loop
+            # over cells (issue (#138) and allow "gh_write" and "gh_readwrite"
```
Thanks, removed.

TeranIvy commented:
I hope I addressed all review comments. I also updated the branch with the latest master which messed up coverage report (again). Testing with compilation works fine for me with Intel 17.0.1 compiler. Back to @rupertford for review.
